### PR TITLE
ranger: simple improvement for range calc for pk.

### PR DESCRIPTION
**Pull request by winoros**
Now we have full ability to calculate the int primary key's range if no cast function occurs.
PTAL @coocood @hanfei1991 @lamxTyler

### CODE REVIEW

File: expression/expression.go [hunk 1]
```diff
@@ -145,6 +145,34 @@ func ComposeDNFCondition(ctx context.Context, conditions ...Expression) Expressi
 	return composeConditionWithBinaryOp(ctx, conditions, ast.LogicOr)
 }
 
+// ExtractDNFConditions extract DNF expression's leaf item.
```
Comment (hanfei1991) [inline expression/expression.go]: extracts

File: expression/expression.go [hunk 2]
```diff
@@ -145,6 +145,34 @@ func ComposeDNFCondition(ctx context.Context, conditions ...Expression) Expressi
...
+// e.g. or(or(a=1, a=2), or(a=3, a=4)), we'll get [a=1, a=2, a=3, a=4].
+func ExtractDNFConditions(DNFCondition *ScalarFunction) []Expression {
```
Comment (hanfei1991) [inline expression/expression.go]: I think DNFItems is more proper.

File: util/ranger/new_refiner.go [hunk 3]
```diff
@@ -132,9 +133,73 @@ func getEQColOffset(expr expression.Expression, cols []*expression.Column) int {
 	return -1
 }
 
+func detachColumnCNFConditions(conditions []expression.Expression, checker *conditionChecker) ([]expression.Expression, []expression.Expression) {
```
Comment (hanfei1991) [inline util/ranger/new_refiner.go]: add some comments

File: util/ranger/new_refiner.go [hunk 4]
```diff
@@ -132,9 +133,73 @@ func getEQColOffset(expr expression.Expression, cols []*expression.Column) int {
...
+	var accessConditions, filterConditions []expression.Expression
+	for _, cond := range conditions {
+		if sf, ok := cond.(*expression.ScalarFunction); ok && sf.FuncName.L == ast.LogicOr {
+			dnfExprs := expression.ExtractDNFConditions(sf)
+			extractedDNFExprs, hasOthers := detachColumnDNFConditions(dnfExprs, checker)
```
Comment (hanfei1991) [inline util/ranger/new_refiner.go]: hasOthers -> hasResidual ?

File: util/ranger/new_refiner.go [hunk 5]
```diff
@@ -132,9 +133,73 @@ func getEQColOffset(expr expression.Expression, cols []*expression.Column) int {
...
+		}
+		accessConditions = append(accessConditions, cond)
+		if checker.shouldReserve {
+			filterConditions = append(filterConditions, cond)
+			checker.shouldReserve = false
+		}
+	}
+	return accessConditions, filterConditions
+}
+
```
Comment (hanfei1991) [inline util/ranger/new_refiner.go]: ditto.

File: expression/expression.go [hunk 6]
```diff
@@ -145,6 +145,34 @@ func ComposeDNFCondition(ctx context.Context, conditions ...Expression) Expressi
...
+		} else {
+			ret = append(ret, arg)
+		}
+	}
+	return ret
+}
+
+// ExtractCNFConditions extract CNF expression's leaf item.
+// e.g. and(and(a>1, a>2), and(a>3, a>4)), we'll get [a>1, a>2, a>3, a>4].
+func ExtractCNFConditions(CNFCondition *ScalarFunction) []Expression {
```
Comment (lamxTyler) [inline expression/expression.go]: Seems like that this function is almost the same with `ExtractDNFConditions`, could we merge them?

File: util/ranger/new_refiner.go [hunk 7]
```diff
@@ -132,9 +133,73 @@ func getEQColOffset(expr expression.Expression, cols []*expression.Column) int {
...
+func detachColumnDNFConditions(conditions []expression.Expression, checker *conditionChecker) ([]expression.Expression, bool) {
```
Comment (lamxTyler) [inline util/ranger/new_refiner.go]: Is this almost the same with `detachColumnCNFConditions`? Could we merge them?

File: util/ranger/range_test.go [hunk 8]
```diff
@@ -174,6 +174,14 @@ func (s *testRangerSuite) TestTableRange(c *C) {
 			exprStr:   "a IS NOT FALSE",
 			resultStr: "[(-inf,-1] [1,+inf)]",
 		},
+		{
```
Comment (hanfei1991) [inline util/ranger/range_test.go]: Can we also output the access and filter conditions ?

**Comment by winoros** [comment]
ok i'll add this in test.

File: util/ranger/range_test.go [hunk 9]
```diff
@@ -63,116 +63,183 @@ func (s *testRangerSuite) TestTableRange(c *C) {
...
+			filterConds: "[]",
+			resultStr:   "[(-inf,-1] [1,+inf)]",
 		},
 		{
-			exprStr:   "a IS NOT FALSE",
-			resultStr: "[(-inf,-1] [1,+inf)]",
+			exprStr:     "a = 1 or a = 3 or a = 4 or (a > 1 and (a = -1 or a = 5))",
+			accessConds: "[or(or(eq(test.t.a, 1), eq(test.t.a, 3)), or(eq(test.t.a, 4), and(gt(test.t.a, 1), or(eq(test.t.a, -1), eq(test.t.a, 5)))))]",
+			filterConds: "[]",
+			resultStr:   "[[1,1] [3,3] [4,4] [5,5]]",
```
Comment (hanfei1991) [inline util/ranger/range_test.go]: Where is -1?

File: util/ranger/range_test.go [hunk 10]
```diff
@@ -63,116 +63,183 @@ func (s *testRangerSuite) TestTableRange(c *C) {
...
-			resultStr: "[[0,0]]",
+			exprStr:     "a IS NOT FALSE",
+			accessConds: "[not(isfalse(test.t.a))]",
+			filterConds: "[]",
+			resultStr:   "[(-inf,-1] [1,+inf)]",
 		},
 		{
-			exprStr:   "a IS NOT FALSE",
-			resultStr: "[(-inf,-1] [1,+inf)]",
+			exprStr:     "a = 1 or a = 3 or a = 4 or (a > 1 and (a = -1 or a = 5))",
```
Comment (hanfei1991) [inline util/ranger/range_test.go]: and a case that `a = 1 or a = 3 or a = 4 or (b > 1 and (a = -1 or a = 5))`

File: util/ranger/new_refiner.go [hunk 11]
```diff
@@ -132,9 +133,77 @@ func getEQColOffset(expr expression.Expression, cols []*expression.Column) int {
...
+			accessConditions = append(accessConditions, cond)
+			if checker.shouldReserve {
+				hasResidualConditions = true
+				checker.shouldReserve = false
+			}
+		} else {
+			return nil, true
+		}
+	}
+	if !hasResidualConditions {
```
Comment (coocood) [inline util/ranger/new_refiner.go]: just 
```
return accessConditions, hasResidualConditions
```
will do

File: expression/expression.go [hunk 12]
```diff
@@ -145,6 +145,30 @@ func ComposeDNFCondition(ctx context.Context, conditions ...Expression) Expressi
...
+		} else {
+			ret = append(ret, arg)
+		}
+	}
+	return ret
+}
+
+// ExtractDNFItems extracts DNF expression's leaf item.
+// e.g. or(or(a=1, a=2), or(a=3, a=4)), we'll get [a=1, a=2, a=3, a=4].
+func ExtractDNFItems(DNFCondition *ScalarFunction) []Expression {
```
Comment (coocood) [inline expression/expression.go]: Maybe `FlattenDNFCondition` is a better name, like `flattenPushDownPlan`.

File: expression/expression.go [hunk 13]
```diff
@@ -145,6 +145,30 @@ func ComposeDNFCondition(ctx context.Context, conditions ...Expression) Expressi
...
+	return extractBinaryOpItems(DNFCondition, ast.LogicOr)
+}
+
+// ExtractCNFItems extracts CNF expression's leaf item.
+// e.g. and(and(a>1, a>2), and(a>3, a>4)), we'll get [a>1, a>2, a>3, a>4].
+func ExtractCNFItems(CNFCondition *ScalarFunction) []Expression {
```
Comment (coocood) [inline expression/expression.go]: Ditto.

File: executor/executor_test.go [hunk 14]
```diff
@@ -2238,3 +2238,13 @@ func (s *testSuite) TestSubqueryInValues(c *C) {
 	tk.MustExec("insert into t (id, name) value ((select gid from t1) ,'asd')")
 	tk.MustQuery("select * from t").Check(testkit.Rows("1 asd"))
 }
+
+func (s *testSuite) TestEnhancedRangeAccess(c *C) {
```
Comment (shenli) [inline executor/executor_test.go]: What's the purpose of this case? Could you comment it?

**Comment by winoros** [comment]
PTAL @coocood

**Comment by hanfei1991** [comment]
LGTM

**Comment by hanfei1991** [comment]
@coocood PTAL

**Comment by lamxTyler** [comment]
@winoros  Please resolve the conflict.

**Pull request merged**